File: 3. Question/CourseSchedule2.cpp
class Solution {
public:
    bool topologicalSort(int n, vector<int> &indegree, vector<vector<int>>& prerequisites){
        vector<vector<int>> adjList(n);
        vector<int> ans;
        for(auto &edges: prerequisites){
            adjList[edges[1]].push_back(edges[0]);
        }

        for(int i=0; i<n; i++){
            for(int x: adjList[i]){
                indegree[x]+=1;
            }
        }

        queue<int> queue;
        for(int i=0; i<n;i++){
            if(indegree[i]==0)
                queue.push(i);
        }
        while(!queue.empty()){
            int front = queue.front();
            queue.pop();
            ans.push_back(front);
            for(int neighbour: adjList[front]){
                indegree[neighbour] --;
                if(indegree[neighbour] == 0){
                    queue.push(neighbour);
                }
            }
        }
        if(ans.size()!=n){
            return false;
        }
        return true;
    }

    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> indegree(numCourses,0);
        return topologicalSort(numCourses, indegree, prerequisites);
    }
    
};

File: 3. Question/CourseSchedule2.py
from collections import deque

def topologicalSort(n, indegree, prerequisites):
    adjList = [[] for _ in range(n+1)]
    ans = []

    for a, b in prerequisites:
        adjList[b].append(a)
    
    for i in range(0, n):
        for x in adjList[i]:
            indegree[x]+=1


    queue = deque()
    for i in range(0, n): 
        if indegree[i] == 0:
            queue.append(i)
    while queue:
        front = queue.popleft()
        ans.append(front)

        for neighbour in adjList[front]:
            indegree[neighbour] -= 1
            if indegree[neighbour] == 0:
                queue.append(neighbour)

    if len(ans) != n:
        return []

    return ans

def findSchedule(numCourses, prerequisites):
    indegree = [0] * (numCourses+1)
    return topologicalSort(numCourses, indegree, prerequisites)
